Failure to search for implicit type tag inside ~ of macro

Both
```scala
class Index[K]
object Index {
  inline def succ[K]: Unit = ~{
    implicit val k = '[K]
    '(new Index[K])
  }
}
```

```scala
class Index[K]
object Index {
  inline def succ[K]: Unit = ~{
    '(new Index[K])
  }
}
```

fail with 
```
Exception in thread "main" java.lang.AssertionError: assertion failed: unresolved symbols: type K when pickling tests/run/tasty-indexed-map/quoted_1.scala
	at scala.Predef$.assert(Predef.scala:219)
	at dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:639)
	at dotty.tools.dotc.core.quoted.PickledQuotes$.pickle(PickledQuotes.scala:98)
	at dotty.tools.dotc.core.quoted.PickledQuotes$.pickleQuote(PickledQuotes.scala:31)
	at dotty.tools.dotc.transform.ReifyQuotes$Reifier.pickleAsTasty$1(ReifyQuotes.scala:401)
	at dotty.tools.dotc.transform.ReifyQuotes$Reifier.pickledQuote(ReifyQuotes.scala:404)
	at dotty.tools.dotc.transform.ReifyQuotes$Reifier.quotation(ReifyQuotes.scala:389)
	....
```

--- a/compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala
+++ b/compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala
@@ -120,7 +120,7 @@ class ReifyQuotes extends MacroTransformWithImplicits with InfoTransformer {
      *
      *  See `isCaptured`
      */
-    val capturers = new mutable.HashMap[Symbol, RefTree => Tree]
+    val capturers = new mutable.HashMap[Symbol, Tree => Tree]
   }
 
   /** The main transformer class
@@ -143,6 +143,9 @@ class ReifyQuotes extends MacroTransformWithImplicits with InfoTransformer {
 
     /** We are in a `~(...)` context that is not shadowed by a nested `'(...)` */
     def inSplice = outer != null && !inQuote
+
+    /** We are not in a `~(...)` or a `'(...)` */
+    def isRoot = outer != null
 
     /** A map from type ref T to expressions of type `quoted.Type[T]`".
      *  These will be turned into splices using `addTags` and represent type variables
@@ -291,6 +294,8 @@ class ReifyQuotes extends MacroTransformWithImplicits with InfoTransformer {
         else i"${sym.name}.this"
       if (!isThis && sym.maybeOwner.isType && !sym.is(Param))
         check(sym.owner, sym.owner.thisType, pos)
+      else if (level == 1 && sym.isType && sym.is(Param) && sym.owner.is(Macro) && outer.isRoot)
+        importedTags(sym.typeRef) = capturers(sym)(ref(sym))
       else if (sym.exists && !sym.isStaticOwner && !levelOK(sym))
         for (errMsg <- tryHeal(tp, pos))
           ctx.error(em"""access to $symStr from wrong staging level:

--- a/compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala
+++ b/compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala
@@ -145,7 +145,7 @@ class ReifyQuotes extends MacroTransformWithImplicits with InfoTransformer {
     def inSplice = outer != null && !inQuote
 
     /** We are not in a `~(...)` or a `'(...)` */
-    def isRoot = outer != null
+    def isRoot = outer == null
 
     /** A map from type ref T to expressions of type `quoted.Type[T]`".
      *  These will be turned into splices using `addTags` and represent type variables
@@ -294,7 +294,7 @@ class ReifyQuotes extends MacroTransformWithImplicits with InfoTransformer {
         else i"${sym.name}.this"
       if (!isThis && sym.maybeOwner.isType && !sym.is(Param))
         check(sym.owner, sym.owner.thisType, pos)
-      else if (level == 1 && sym.isType && sym.is(Param) && sym.owner.is(Macro) && outer.isRoot)
+      else if (level == 1 && sym.isType && sym.is(Param) && sym.owner.is(Macro) && !outer.isRoot)
         importedTags(sym.typeRef) = capturers(sym)(ref(sym))
       else if (sym.exists && !sym.isStaticOwner && !levelOK(sym))
         for (errMsg <- tryHeal(tp, pos))

--- a/tests/pos/i4493-b.scala
+++ b/tests/pos/i4493-b.scala
@@ -1,0 +1,7 @@
+class Index[K]
+object Index {
+  inline def succ[K](x: K): Unit = ~{
+    implicit val t: quoted.Type[K] = '[K]
+    '(new Index[K])
+  }
+}

--- a/tests/pos/i4493-c.scala
+++ b/tests/pos/i4493-c.scala
@@ -1,0 +1,6 @@
+class Index[K]
+object Index {
+  inline def succ[K]: Unit = ~{
+    '(new Index[K])
+  }
+}

--- a/tests/pos/i4493.scala
+++ b/tests/pos/i4493.scala
@@ -1,0 +1,7 @@
+class Index[K]
+object Index {
+  inline def succ[K]: Unit = ~{
+    implicit val t: quoted.Type[K] = '[K]
+    '(new Index[K])
+  }
+}

--- a/tests/pos/i4514.scala
+++ b/tests/pos/i4514.scala
@@ -1,0 +1,4 @@
+object Foo {
+  inline def foo[X](x: X): Unit = ~fooImpl('(x))
+  def fooImpl[X: quoted.Type](x: X): quoted.Expr[Unit] = '()
+}

